### PR TITLE
`<random>`: Fixes subtract_with_carry_engine io

### LINKED ISSUE #437: <random>: Cannot save/restore a ranlux48 object in a stream
**Describe the bug**
It appears that the textual representation of ranlux48 is incorrect

**Command-line test case**
STL version (git commit or Visual Studio version): Visual Studio 2019 version 16.4 (but goes back effectively forever)

```
C:\Users\bion\Desktop>type test.cpp
#include <iostream>
#include <random>
#include <sstream>

template <typename RND> void check(const char *name) {
  RND g1, g2;
  std::cout << std::boolalpha << name << ": " << (g1 == g2);
  std::stringstream s;
  s << g1;
  s >> g1;
  std::cout << " " << (g1 == g2) << std::endl;
}

int main() {
  check<std::ranlux48>("ranlux48");
  check<std::ranlux24>("ranlux24");
  check<std::mt19937>("mt19937");
}

C:\Users\bion\Desktop>cl /EHsc /W4 /WX /std:c++latest /nologo .\test.cpp
test.cpp

C:\Users\bion\Desktop>.\test.exe
ranlux48: true false
ranlux24: true true
mt19937: true true

C:\Users\bion\Desktop>
```

**Expected behavior**
ranlux48: true true 
ranlux24: true true 
mt19937: true true

This issue is a dual of Microsoft-internal VSO-104809 and DevCom-249085

### LINKED ISSUE #442: <random>: subtract_with_carry_engine textual representation is incorrect
**Describe the bug**
In our `subtract_with_carry_engine`, if the `result_type` is 64 bits, we are still emitting 32 bit values. The standard says:

>The textual representation consists of the values of Xi−r, . . . ,Xi−1, in that order, followed by c.

**Command-line test case**
STL version (git commit or Visual Studio version):

```
C:\Users\bion\Desktop>type test.cpp
#include <assert.h>
#include <iostream>
#include <random>
#include <sstream>
#include <string_view>

using namespace std::string_view_literals;

constexpr auto expected =
    "3058990831114416762 8271088423883502182 8539773296756973456 "
    "7389019366887995718 570256334847185029 3716502990755565555 "
    "1524729574385026041 3137027355700498078 462318917886328761 "
    "4995455229142172915 4449332696079976531 1322748335426821320 "
    "1481389978760661022 8563059522958733270 3849224045900943705 "
    "4972782340528344279 5835416012924508119 1706253864337811739 "
    "5350051069717426128 2939050575112930398 2997396533930240412 "
    "2702983745895389727 6734488758036272684 2336722746921380976 0 0"sv;

int main() {
  std::subtract_with_carry_engine<unsigned long long, 64, 10, 24> ull_swc;
  ull_swc.seed(0x12341234'00000000ULL);
  std::ostringstream oss;
  oss << ull_swc;
  const auto actual = std::move(oss).str();
  std::cout << "Expected: " << expected << "\n"sv;
  std::cout << "Actual: " << actual << "\n"sv;
  assert(expected == actual);
}

C:\Users\bion\Desktop>cl /EHsc /W4 /WX /std:c++latest /nologo .\test.cpp
test.cpp

C:\Users\bion\Desktop>.\test.exe
Expected: 3058990831114416762 8271088423883502182 8539773296756973456 7389019366887995718 570256334847185029 3716502990755565555 1524729574385026041 3137027355700498078 462318917886328761 4995455229142172915 4449332696079976531 1322748335426821320 1481389978760661022 8563059522958733270 3849224045900943705 4972782340528344279 5835416012924508119 1706253864337811739 5350051069717426128 2939050575112930398 2997396533930240412 2702983745895389727 6734488758036272684 2336722746921380976 0 0
Actual: 712226804 599814778 1925762841 1936454246 1988320913 1469112208 1720390135 701970758 132773149 2105249925 865315783 2057932787 355003768 868254713 730396098 1664487070 107642011 965656505 1163095056 1482884339 1035940995 1969277011 307976346 1415240904 344912982 1104824350 1993742660 1618685910 896217312 751915353 1157816113 411503831 1358663666 1191040983 397268185 2021533979 1245655834 615821264 684301037 579044446 697885764 1206266268 629337445 1472191007 1567995352 916264492 544060661 886238320 0
Assertion failed: expected == actual, file .\test.cpp, line 27
```

Also tracked by DevCom-249085 and Microsoft-internal VSO-104809 / [AB#104809](https://devdiv.visualstudio.com/0bdbc590-a062-4c3f-b0f6-9383f67865ee/_workitems/edit/104809).

### PR DESCRIPTION
**Pull request by yuanhongzhao**
- Fixes #442. Also fixes #437, which was a slightly different issue. I maintained the textual representation of `tr1::subtract_with_carry`, breaking the state into 32-bit words, but the streaming operators should be consistent now.
 - Implements part of LWG-3519 by making `subtract_with_carry_engine` streaming operators hidden `friend`s. `tr1` classes continue to go through a free function taking an `_Swc_base&`.
 - Improves conformance of `subtract_with_carry_engine` streaming operators [[tab:rand.req.eng]](http://eel.is/c++draft/tab:rand.req.eng), specifically the handling of the stream's flags and leaving the state unchanged if reading fails.
 - The now-unskipped libc++ tests cover the streaming operators and textual representation, so I only added coverage for the stream flags.

### CODE REVIEW

File: tests/std/tests/GH_000442_rand_swc_io/test.cpp [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        "96783156950859",
+        "238199764491708",
+        "34339434557790",
+        "155299155394531",
+        "29014415493780",
+        "209265474179052",
+        "263777435457028",
+        "0",
+    };
+    constexpr auto state_size = sizeof(state_ref) / sizeof(state_ref[0]);
```
Comment (StephanTLavavej) [inline tests/std/tests/GH_000442_rand_swc_io/test.cpp:28]: Suggestion: `size()` from `<iterator>` would be more convenient/modern.

File: tests/std/tests/GH_000442_rand_swc_io/test.cpp [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    stringstream sstr(state_str);
+
+    size_t idx = 0;
```
Comment (StephanTLavavej) [inline tests/std/tests/GH_000442_rand_swc_io/test.cpp:31]: Technically should include `<cstddef>`.

File: tests/std/tests/GH_000442_rand_swc_io/test.cpp [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    string word;
+    while (sstr && idx < state_size) {
+        sstr >> word;
+        assert(word == state_ref[idx]);
+        ++idx;
+    }
+
+    assert(sstr.rdstate() == ios_base::eofbit && idx == state_size);
```
Comment (StephanTLavavej) [inline tests/std/tests/GH_000442_rand_swc_io/test.cpp:39]: Suggestion: When tests fail, `assert(X && Y)` doesn't indicate which one failed. Separately asserting X and Y is generally worth the extra line.

**Review by StephanTLavavej — APPROVED** [review]

**Review by CaseyCarter — APPROVED** [review]

**Comment by CaseyCarter** [comment]
Thank you for the bug fix!

**Pull request merged**